src/FsAutoComplete/LspServers/ProjectWorkspace.fs

```diff
@@ -0,0 +1,298 @@
+namespace FsAutoComplete.ProjectWorkspace
+
+open System
+
+module Snapshots =
+  open System
+  open FsAutoComplete
+  open System.Threading
+  open FSharp.UMX
+  open System.Threading.Tasks
+  open Ionide.ProjInfo.Types
+  open FSharp.Compiler.CodeAnalysis.ProjectSnapshot
+  open System.IO
+  open FSharp.Compiler.CodeAnalysis
+  open FSharp.Data.Adaptive
+  open FSharp.Compiler.Text
+  open FsAutoComplete.Adaptive
+  open Ionide.ProjInfo.Logging
+  open System.Collections.Generic
+
+  let rec logger = LogProvider.getLoggerByQuotation <@ logger @>
+
+  let private loadFromDotnetDll (p: ProjectOptions) : FSharpReferencedProjectSnapshot =
+    /// because only a successful compilation will be written to a DLL, we can rely on
+    /// the file metadata for things like write times
+    let projectFile = FileInfo p.TargetPath
+
+    let getStamp () =
+      projectFile.Refresh()
+      projectFile.LastWriteTimeUtc
+
+    let getStream (_ctok: System.Threading.CancellationToken) =
+      try
+        projectFile.OpenRead() :> Stream |> Some
+      with _ ->
+        None
+
+    let delayedReader = DelayedILModuleReader(p.TargetPath, getStream)
+
+    ProjectSnapshot.FSharpReferencedProjectSnapshot.PEReference(getStamp, delayedReader)
+
+  let makeAdaptiveFCSSnapshot
+    projectFileName
+    projectId
+    sourceFiles
+    referencePaths
+    otherOptions
+    referencedProjects
+    isIncompleteTypeCheckEnvironment
+    useScriptResolutionRules
+    loadTime
+    unresolvedReferences
+    originalLoadReferences
+    =
+    aval {
+      let! projectFileName = projectFileName
+      and! projectId = projectId
+      and! sourceFiles = sourceFiles
+      and! referencePaths = referencePaths
+      and! otherOptions = otherOptions
+      and! referencedProjects = referencedProjects
+      and! isIncompleteTypeCheckEnvironment = isIncompleteTypeCheckEnvironment
+      and! useScriptResolutionRules = useScriptResolutionRules
+      and! loadTime = loadTime
+      and! unresolvedReferences = unresolvedReferences
+      and! originalLoadReferences = originalLoadReferences
```

Comment:
Note, any changes to any of these could recreate a snapshot. In practice only a few of these change